enumeration/enumerator/cloud_enumerator.go

```diff
@@ -0,0 +1,266 @@
+package enumerator
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/snyk/driftctl/enumeration"
+
+	"github.com/sirupsen/logrus"
+	"github.com/snyk/driftctl/enumeration/alerter"
+	"github.com/snyk/driftctl/enumeration/parallel"
+	"github.com/snyk/driftctl/enumeration/remote"
+	"github.com/snyk/driftctl/enumeration/remote/common"
+	"github.com/snyk/driftctl/enumeration/resource"
+	"github.com/snyk/driftctl/enumeration/terraform"
+)
+
+type CloudEnumerator struct {
+	alerter              *sliceAlerter
+	progress             enumeration.ProgressCounter
+	remoteLibrary        *common.RemoteLibrary
+	providerLibrary      *terraform.ProviderLibrary
+	enumeratorRunner     *parallel.ParallelRunner
+	detailsFetcherRunner *parallel.ParallelRunner
+	to                   string
+}
+
+type cloudEnumeratorBuilder struct {
+	cloud           string
+	providerVersion string
+	configDirectory string
+}
+
+// WithCloud Choose which cloud to use for enumeration and refresh
+// TODO could be inferred with types listed
+func (b *cloudEnumeratorBuilder) WithCloud(cloud string) *cloudEnumeratorBuilder {
+	b.cloud = cloud
+	return b
+}
+
+// WithProviderVersion optionally choose the provider version used for refresh
+func (b *cloudEnumeratorBuilder) WithProviderVersion(providerVersion string) *cloudEnumeratorBuilder {
+	b.providerVersion = providerVersion
+	return b
+}
```

Comment:
Those two parameters are required (we could make the cloud one optional if we can detect it from the resource type list)

Having them in the builder pattern allow use to create an invalid struct if we do that:

```go
NewCloudEnumerator().Build()
```

Even for the enumeration without refresh usecase it will fail since we always call the `init()` func in the `Build` method.

I think we should add a check in Build method to avoid calling `init()` if we don't have both cloud and version parameters set.

Or maybe also create a `NewCloudEnumerator().WithRefresh("aws", "4.4.0").Build()` could be more explicit ?

Let's also add a check in the refresh to prevent the execution if we instanciate a CloudEnumerator without refresh capability.

Happy to pair on that BTW

Reply:
> Those two parameters are required (we could make the cloud one optional if we can detect it from the resource type list)

version is not required. the provider init will take a default one when it's not provided (https://github.com/snyk/driftctl/blob/cf9dba0712c76a593857684fa9d7b355140c29c3/enumeration/remote/aws/provider.go#L51)
I could check that the cloud is not empty but it should already return an error from init when the cloud is empty or not supported.

>Let's also add a check in the refresh to prevent the execution if we instanciate a CloudEnumerator without refresh capability.

For now there is no real mean to instantiate that as I need to call activate (and it create provider instance) to instantiate the enumerators. Hence I'm not sure we need such a check until we change that ?